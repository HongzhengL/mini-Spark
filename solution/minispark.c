#include "minispark.h"

#include <bits/time.h>
#include <pthread.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "list.h"
#include "thread_pool.h"

#define LIST_INIT_CAPACITY 10
#define CONTENT_INIT_CAPACITY 1024

const char *LOG_FILE = "metrics.log";

MetricQueue *metric_queue;

pthread_t metric_thread;

// Working with metrics...
// Recording the current time in a `struct timespec`:
//    clock_gettime(CLOCK_MONOTONIC, &metric->created);
// Getting the elapsed time in microseconds between two timespecs:
//    duration = TIME_DIFF_MICROS(metric->created, metric->scheduled);
// Use `print_formatted_metric(...)` to write a metric to the logfile.
void print_formatted_metric(TaskMetric *metric, FILE *fp) {
    fprintf(fp,
            "RDD %p Part %d Trans %d -- creation %10jd.%06ld, scheduled "
            "%10jd.%06ld, execution (usec) %ld\n",
            metric->rdd, metric->pnum, metric->rdd->trans,
            metric->created.tv_sec, metric->created.tv_nsec / 1000,
            metric->scheduled.tv_sec, metric->scheduled.tv_nsec / 1000,
            metric->duration);
}

int max(int a, int b) {
    return a > b ? a : b;
}

RDD *create_rdd(int numdeps, Transform t, void *fn, ...) {
    RDD *rdd = malloc(sizeof(RDD));
    if (rdd == NULL) {
        printf("error mallocing new rdd\n");
        exit(1);
    }

    va_list args;
    va_start(args, fn);

    int maxpartitions = 0;
    for (int i = 0; i < numdeps; i++) {
        RDD *dep = va_arg(args, RDD *);
        rdd->dependencies[i] = dep;
        if (dep) maxpartitions = max(maxpartitions, dep->numpartitions);
    }
    va_end(args);

    rdd->numdependencies = numdeps;
    rdd->trans = t;
    rdd->fn = fn;
    rdd->partitions = NULL;
    rdd->numpartitions = maxpartitions;
    rdd->numComputed = 0;
    pthread_mutex_init(&(rdd->partitionListLock), NULL);
    return rdd;
}

/* RDD constructors */
RDD *map(RDD *dep, Mapper fn) {
    // fprintf(stderr, "from minispark.c@66: %p\n",
    // get_nth_elem(dep->partitions, 0));
    return create_rdd(1, MAP, fn, dep);
}

RDD *filter(RDD *dep, Filter fn, void *ctx) {
    RDD *rdd = create_rdd(1, FILTER, fn, dep);
    rdd->ctx = ctx;
    return rdd;
}

RDD *partitionBy(RDD *dep, Partitioner fn, int numpartitions, void *ctx) {
    RDD *rdd = create_rdd(1, PARTITIONBY, fn, dep);
    rdd->partitions = list_init(numpartitions);
    rdd->numpartitions = numpartitions;
    rdd->ctx = ctx;
    return rdd;
}

RDD *join(RDD *dep1, RDD *dep2, Joiner fn, void *ctx) {
    RDD *rdd = create_rdd(2, JOIN, fn, dep1, dep2);
    rdd->ctx = ctx;
    return rdd;
}

/* A special mapper */
void *identity(void *arg) {
    return arg;
}

/* Special RDD constructor.
 * By convention, this is how we read from input files. */
RDD *RDDFromFiles(char **filenames, int numfiles) {
    RDD *rdd = malloc(sizeof(RDD));
    rdd->partitions = list_init(numfiles);

    for (int i = 0; i < numfiles; i++) {
        FILE *fp = fopen(filenames[i], "r");
        if (fp == NULL) {
            perror("fopen");
            exit(1);
        }
        list_add_elem(rdd->partitions, fp);
    }

    rdd->numdependencies = 0;
    rdd->trans = FILE_BACKED;
    rdd->fn = (void *)identity;
    rdd->numpartitions = numfiles;
    rdd->numComputed = 0;
    pthread_mutex_init(&(rdd->partitionListLock), NULL);
    return rdd;
}

static List *get_tasks_list(RDD *rdd) {
    List *queue = list_init(LIST_INIT_CAPACITY);
    List *rdd_list = list_init(LIST_INIT_CAPACITY);

    list_add_elem(queue, rdd);
    list_add_elem(rdd_list, rdd);
    while (get_size(queue) != 0) {
        RDD *curr = list_remove_front(queue);
        for (int i = 0; i < curr->numdependencies; ++i) {
            list_add_elem(queue, curr->dependencies[i]);
            list_add_elem(rdd_list, curr->dependencies[i]);
        }
    }

    List *rev = list_reverse(rdd_list);
    free_list(rdd_list);

    return rev;
}

void execute(RDD *rdd) {
    List *rdds = get_tasks_list(rdd);
    seek_to_start(rdds);
    RDD *rdd_ptr = NULL;
    while ((rdd_ptr = next(rdds)) != NULL) {
        for (int i = 0; i < rdd_ptr->numpartitions; ++i) {
            Task *task = (Task *)malloc(sizeof(Task));
            TaskMetric *metric = (TaskMetric *)malloc(sizeof(TaskMetric));
            task->rdd = rdd_ptr;
            task->pnum = i;
            task->metric = metric;
            metric->pnum = i;
            metric->rdd = rdd_ptr;
            clock_gettime(CLOCK_MONOTONIC, &metric->created);
            thread_pool_submit(task);
        }
    }
}

void *metric_thread_func(void *arg) {
    FILE *fp = (FILE *)arg;
    while (1) {
        pthread_mutex_lock(&metric_queue->queue_lock);
        while (metric_queue->queue->size == 0) {
            pthread_cond_wait(&metric_queue->queue_not_empty,
                              &metric_queue->queue_lock);
        }
        TaskMetric *metric =
            (TaskMetric *)list_remove_front(metric_queue->queue);
        pthread_cond_signal(&metric_queue->queue_not_full);
        pthread_mutex_unlock(&metric_queue->queue_lock);
        if (metric->rdd == NULL) {
            fclose(fp);
            // free(metric);
            break;
        } else {
            print_formatted_metric(metric, fp);
            // free(metric);
        }
    }
    return NULL;
}

void MS_Run() {
    thread_pool_init(get_num_threads());
    metric_queue = (MetricQueue *)malloc(sizeof(MetricQueue));
    metric_queue->queue = list_init(METRIC_QUEUE_CAPACITY);
    pthread_mutex_init(&metric_queue->queue_lock, NULL);
    pthread_cond_init(&metric_queue->queue_not_empty, NULL);
    pthread_cond_init(&metric_queue->queue_not_full, NULL);

    FILE *fp = fopen(LOG_FILE, "w");
    if (fp == NULL) {
        perror("fopen");
        exit(EXIT_FAILURE);
    }
    if (pthread_create(&metric_thread, NULL, metric_thread_func, fp) != 0) {
        perror("pthread_create");
        fclose(fp);
        exit(EXIT_FAILURE);
    }
}

void MS_TearDown() {
    thread_pool_destroy();
    TaskMetric *metric = (TaskMetric *)malloc(sizeof(TaskMetric));
    metric->rdd = NULL;
    pthread_mutex_lock(&metric_queue->queue_lock);
    while (metric_queue->queue->size == METRIC_QUEUE_CAPACITY) {
        pthread_cond_wait(&metric_queue->queue_not_full,
                          &metric_queue->queue_lock);
    }
    list_add_elem(metric_queue->queue, metric);
    pthread_cond_signal(&metric_queue->queue_not_empty);
    pthread_mutex_unlock(&metric_queue->queue_lock);
    pthread_join(metric_thread, NULL);
}

static void free_RDD_resource(RDD *rdd) {
    List *curr = NULL;
    seek_to_start(rdd->partitions);
    while ((curr = (List *)next(rdd->partitions)) != NULL) {
        void *item = NULL;
        seek_to_start(curr);
        while ((item = next(curr)) != NULL) {
            // free(item);
        }
        free_list(curr);
    }
    free_list(rdd->partitions);
}

static void free_All_RDD(RDD *rdd) {
    List *rdds = get_tasks_list(rdd);
    seek_to_start(rdds);
    RDD *rdd_ptr = NULL;
    while ((rdd_ptr = next(rdds)) != NULL) {
        if (rdd_ptr->numdependencies != 0) {
            free_RDD_resource(rdd_ptr);
        }
        free(rdd_ptr);
    }
    free_list(rdds);
}

int count(RDD *rdd) {
    execute(rdd);
    thread_pool_wait();

    int count = 0;
    List *curr = NULL;
    seek_to_start(rdd->partitions);
    while ((curr = (List *)next(rdd->partitions)) != NULL) {
        count += curr->size;
    }

    // free_All_RDD(rdd);

    return count;
}

void print(RDD *rdd, Printer p) {
    // fprintf(stderr, "from minispark.c@240: %p\n",
    // get_nth_elem(rdd->dependencies[0]->partitions, 0));
    execute(rdd);
    // fprintf(stderr, "from minispark.c@242: %p\n",
    // get_nth_elem(rdd->dependencies[0]->partitions, 0));
    thread_pool_wait();

    // print all the items in rdd
    // aka... `p(item)` for all items in rdd
    List *curr = NULL;
    seek_to_start(rdd->partitions);
    while ((curr = (List *)next(rdd->partitions)) != NULL) {
        void *item = NULL;
        seek_to_start(curr);
        while ((item = next(curr)) != NULL) {
            p(item);
        }
    }

    // free_All_RDD(rdd);
}
