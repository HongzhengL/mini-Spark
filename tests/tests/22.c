#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <sched.h>
#include <stdarg.h>
#include <unistd.h>
#include "lib.h"
#include "minispark.h"
#include <string.h>
#define FILENAMESIZE 100
#define GRANULARITY 8

int RowDividable(void* arg, void* needle) {
    
    struct row* row = (struct row*)arg;
    if (atoi(row->cols[0])%128) {
        return 0; //should be very rare
    }
    return 1;
}


// --- Function to read the complete file using mmap ---
// This function opens the file, obtains its size, mmaps it, then copies its content
// into a dynamically allocated buffer (with a null terminator) which is returned.
char *read_serialized_file(const char *filename) {
    int fd = open(filename, O_RDONLY);
    if (fd < 0) {
        perror("open");
        return NULL;
    }
    
    struct stat st;
    if (fstat(fd, &st) < 0) {
        perror("fstat");
        close(fd);
        return NULL;
    }
    
    size_t filesize = st.st_size;
    if (filesize == 0) {
        fprintf(stderr, "Error: File is empty.\n");
        close(fd);
        return NULL;
    }
    
    // Memory-map the file for efficient reading.
    char *mapped = mmap(NULL, filesize, PROT_READ, MAP_PRIVATE, fd, 0);
    if (mapped == MAP_FAILED) {
        perror("mmap");
        close(fd);
        return NULL;
    }
    
    // Allocate a buffer with one extra byte for the null terminator.
    char *buffer = malloc(filesize + 1);
    if (buffer == NULL) {
        perror("malloc");
        munmap(mapped, filesize);
        close(fd);
        return NULL;
    }
    
    memcpy(buffer, mapped, filesize);
    buffer[filesize] = '\0';  // Ensure the data is null-terminated.
    
    munmap(mapped, filesize);
    close(fd);
    return buffer;
}

struct sumjoin_ctx sctx = {0,1};
struct colpart_ctx pctx = {0};

// --- Helper: recursively deserialize the RDD tree ---
// 'tokens' is an array of token strings obtained from the serialized RDD.
// 'idx' points to the current token index (passed by reference so that it is updated).
RDD* deserialize_rdd_helper(char** tokens, int *idx) {
    if (!tokens[*idx]) {
        return NULL;
    }
    // Read the transformation number.
    int trans = atoi(tokens[*idx]);
    (*idx)++;
    
    if (trans == 0) {
        // For transformation 0 (map) the serializer printed two extra fields (backed and numpartitions)
        // that we don’t actually use to reconstruct the function chain.
        // Skip the next two tokens.
        int filenum = atoi(tokens[*idx]);
        (*idx)++;
        (*idx)++;  // skip the 'numpartitions' token
        
        // Build the map chain as specified:
        // map(map(RDDFromFiles("filename", 1), GetLines), SplitCols)
        
        char* filelist[GRANULARITY];
        for (int i = 0; i< GRANULARITY; i++) {
            filelist[i] = calloc(FILENAMESIZE, 1);
            sprintf(filelist[i],  "./test_files/hiddenvals%d.txt", ((filenum)/(GRANULARITY))*GRANULARITY + (filenum+i)%GRANULARITY);
        }

        RDD* source = RDDFromFiles(filelist, GRANULARITY);
        RDD* inner = map(source, GetLines);
        return map(inner, SplitCols);
    }
    else if (trans == 1) {
        // For filter: deserialize its single dependency.
        RDD* dep = deserialize_rdd_helper(tokens, idx);
        return filter(dep, RowDividable, NULL); // literal "0" as the context
    }
    else if (trans == 2) {
  
        // For join: deserialize two dependencies.
        RDD* dep0 = deserialize_rdd_helper(tokens, idx);
        RDD* dep1 = deserialize_rdd_helper(tokens, idx);
        return join(dep0, dep1, SumJoin, (void*)&sctx);
    }
    else if (trans == 3) {
        // For partitionBy: the next token holds the number of partitions.
        int numPartitions = atoi(tokens[*idx]);
        (*idx)++;
        RDD* dep = deserialize_rdd_helper(tokens, idx);
        return partitionBy(dep, ColumnHashPartitioner, numPartitions, &pctx);
    }
    else {
        // Unrecognized transformation.
        fprintf(stderr, "Error: Unrecognized transformation code %d\n", trans);
        return NULL;
    }
}

// --- Main deserialization function ---
// This function tokenizes the input serialized string (assumed to be whitespace delimited)
// and calls the helper to recursively build the RDD tree.
RDD* deserialize_rdd(const char* serialized) {
    // Make a copy so that strtok doesn’t modify the caller’s string.
    char* str = strdup(serialized);
    if (str == NULL) {
        perror("strdup");
        return NULL;
    }
    
    // First, count tokens (this simple approach assumes tokens are separated by whitespace)
    int tokenCap = 16;
    int tokenCount = 0;
    char** tokens = malloc(tokenCap * sizeof(char*));
    if (tokens == NULL) {
        free(str);
        perror("malloc");
        return NULL;
    }
    char *saveptr1;
    // Tokenize using strtok (tokens will be separated by spaces, tabs, or newlines).
    char* token = strtok_r(str, " \t\n", &saveptr1);
    while (token != NULL) {
        if (tokenCount >= tokenCap) {
            tokenCap *= 2;
            tokens = realloc(tokens, tokenCap * sizeof(char*));
            if (tokens == NULL) {
                free(str);
                perror("realloc");
                return NULL;
            }
        }
        tokens[tokenCount++] = token;
        token = strtok_r(NULL, " \t\n", &saveptr1);
    }
    
    // Use an index to keep track of parsing progress.
    int idx = 0;
    RDD* result = deserialize_rdd_helper(tokens, &idx);
    
    // Cleanup
    free(tokens);
    free(str);
    return result;
}


int main(int argc, char** argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <serialized_file>\n", argv[0]);
        return EXIT_FAILURE;
    }
    
    // Read serialized RDD data from the file.
    char *serialized_data = read_serialized_file(argv[1]);
    if (serialized_data == NULL) {
        fprintf(stderr, "Failed to read serialized data from file.\n");
        return EXIT_FAILURE;
    }
    
    // Reconstruct the RDD from the serialized data.
    RDD* rdd = deserialize_rdd(serialized_data);
    if (rdd == NULL) {
        fprintf(stderr, "Deserialization failed.\n");
        free(serialized_data);
        return EXIT_FAILURE;
    }

    MS_Run();
    print(rdd, RowPrinter);
    MS_TearDown();

    free(serialized_data);
    return EXIT_SUCCESS;
}