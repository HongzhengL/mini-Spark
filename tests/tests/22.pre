python3 tests/22.py > tests/22.rdd