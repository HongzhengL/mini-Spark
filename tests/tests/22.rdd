2 3 4 3 3 3 2 2 3 6 3 1 1 3 1 2 3 6 1 1 1 2 3 5 3 1 2 3 8 3 1 2 3 5 0 294 1 3 5 2 3 4 0 840 74 3 4 0 106 81 3 8 2 3 3 1 3 1 3 1 0 509 19 3 3 2 3 1 0 260 41 3 1 0 955 115 3 5 3 1 2 3 7 2 3 6 3 1 1 0 467 1 3 6 1 0 687 2 3 7 1 3 1 2 3 4 1 2 3 2 0 259 121 3 2 0 117 113 3 4 1 0 412 3 3 6 2 3 3 3 1 2 3 4 3 1 0 506 124 3 4 2 3 4 1 2 3 8 2 3 6 0 1013 90 3 6 0 56 77 3 8 0 330 96 3 4 0 1004 5 3 3 2 3 5 1 3 1 3 1 2 3 7 1 1 0 451 2 3 7 2 3 3 0 223 109 3 3 3 1 0 779 6 3 5 3 1 3 1 2 3 1 0 276 50 3 1 0 560 48 3 6 1 2 3 6 2 3 7 3 3 1 1 2 3 6 1 2 3 3 0 678 31 3 3 2 3 8 2 3 5 0 613 3 3 5 0 539 125 3 8 0 221 44 3 6 3 3 3 3 3 3 3 3 2 3 3 0 790 3 3 3 3 3 2 3 1 1 0 733 6 3 1 0 282 88 3 7 3 3 2 3 2 2 3 2 2 3 6 0 765 62 3 6 0 321 66 3 2 2 3 4 0 712 107 3 4 3 3 0 564 119 3 2 2 3 5 2 3 7 2 3 3 0 840 3 3 3 0 32 26 3 7 1 0 935 6 3 5 3 3 2 3 1 0 616 49 3 1 0 732 50 3 6 3 3 3 3 1 1 3 3 3 3 1 2 3 6 3 3 3 3 2 3 7 1 1 1 2 3 8 0 746 69 3 8 0 239 116 3 7 1 3 3 0 934 3 3 6 2 3 1 2 3 4 2 3 8 0 727 112 3 8 0 444 75 3 4 2 3 5 2 3 1 0 641 25 3 1 0 338 5 3 5 0 545 5 3 1 1 1 1 3 3 1 0 747 6 3 4 1 2 3 1 1 2 3 2 2 3 5 1 3 1 1 1 2 3 2 1 3 1 3 1 0 470 6 3 2 2 3 3 1 1 0 364 4 3 3 3 1 0 928 41 3 5 1 2 3 1 2 3 7 2 3 4 0 751 96 3 4 0 213 101 3 7 2 3 1 0 300 110 3 1 0 406 78 3 1 2 3 7 0 188 1 3 7 3 1 0 419 54 3 2 1 2 3 6 2 3 7 3 1 3 1 1 2 3 2 1 3 1 0 276 1 3 2 3 1 0 611 91 3 7 3 1 0 698 1 3 6 3 1 1 2 3 5 2 3 7 0 141 101 3 7 0 584 11 3 5 3 1 1 0 829 1 3 1 3 1 3 1 3 1 3 1 3 1 2 3 4 3 1 2 3 8 2 3 8 1 2 3 7 0 142 59 3 7 0 654 44 3 8 2 3 4 3 1 3 1 3 1 0 104 55 3 4 0 432 70 3 8 2 3 6 3 1 2 3 7 2 3 6 0 920 61 3 6 3 1 1 0 979 2 3 7 3 1 2 3 7 1 0 383 6 3 7 1 0 148 4 3 6 2 3 4 3 1 2 3 3 1 2 3 5 0 313 40 3 5 0 283 53 3 3 0 565 122 3 4 1 3 1 2 3 3 0 803 76 3 3 0 822 104 3 4 3 1 1 3 1 2 3 4 1 1 2 3 2 2 3 5 0 425 101 3 5 2 3 3 1 0 937 6 3 3 3 1 2 3 3 0 559 128 3 3 0 43 103 3 2 2 3 3 2 3 1 0 787 66 3 1 0 596 1 3 3 2 3 8 1 2 3 3 2 3 1 0 398 50 3 1 2 3 6 0 631 100 3 6 0 665 36 3 3 0 630 13 3 8 0 438 101 3 4 3 1 1 3 1 2 3 7 3 1 1 3 1 0 322 1 3 7 1 1 3 1 0 85 1 
