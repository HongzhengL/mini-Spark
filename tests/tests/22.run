./tests/22.tmp ./tests/22.rdd