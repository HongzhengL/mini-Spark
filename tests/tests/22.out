1280	462748
896	422611
3840	423974
7040	470142
4096	387280
6656	410559
1536	442374
384	415741
2944	397931
1024	363491
6144	444869
1792	391752
7808	515793
2432	397963
7552	377178
128	446598
5248	435893
1920	483578
7296	454811
0	438959
5120	485097
2176	470696
4736	430551
4480	421967
3584	402637
4224	433668
3072	431534
512	474790
2688	411198
5632	445107
3328	415001
4992	442674
256	400213
3200	422734
5376	433078
7936	340559
2816	461666
6016	438316
7680	435805
2560	411535
2304	441534
7424	435068
6784	437117
1664	424711
3712	454181
5888	459159
768	380341
1408	496562
6528	420510
6272	425189
1152	406411
5760	456594
640	482700
6400	423253
3456	444267
5504	430278
4864	473133
8064	517087
2048	443668
7168	433575
4608	442060
4352	493816
6912	510940
3968	482694
